sgl-pdlb/src/server.rs

```diff
@@ -0,0 +1,340 @@
+}
+
+impl LBState {
+    pub fn new(lb_config: LBConfig) -> anyhow::Result<Self> {
+        let client = reqwest::Client::builder()
+            .timeout(std::time::Duration::from_secs(lb_config.timeout))
+            .build()?;
+        let policy = match lb_config.policy.as_str() {
+            "random" => LBPolicy::Random,
+            "po2" => LBPolicy::PowerOfTwo,
+            _ => anyhow::bail!("Invalid policy"),
+        };
+        let prefill_servers = lb_config
+            .prefill_infos
+            .into_iter()
+            .map(|(url, port)| EngineInfo::new_prefill(url, port))
+            .collect();
+        let decode_servers = lb_config
+            .decode_infos
+            .into_iter()
+            .map(|url| EngineInfo::new_decode(url))
+            .collect();
+        let lb = StrategyLB::new(policy, prefill_servers, decode_servers);
+        Ok(Self {
+            strategy_lb: lb,
+            client,
+            log_interval: lb_config.log_interval,
+        })
+    }
+
+    async fn route_one(
+        &self,
+        engine_info: &EngineInfo,
+        method: Method,
+        api_path: &str,
+        request: Option<&serde_json::Value>,
+        stream: bool,
+    ) -> Result<ProxyResponse, actix_web::Error> {
+        let url = engine_info.api_path(api_path);
+        let request = request.unwrap_or(&serde_json::Value::Null);
+        let task = self.client.request(method, url).json(request).send();
+        let resp = task.await.map_err(actix_web::error::ErrorBadGateway)?;
+        // FIXME: handle error status code (map status code to error)
+        let status = resp.status();
+        let body = if stream {
+            let resp_stream = resp.bytes_stream().map(|r| {
+                r.map_err(actix_web::error::ErrorBadGateway)
+                    .map(Bytes::from)
+            });
+            ProxyResponseBody::Stream(Box::pin(resp_stream))
+        } else {
+            let body = resp
+                .bytes()
+                .await
+                .map_err(actix_web::error::ErrorBadGateway)?;
+            ProxyResponseBody::Full(body)
+        };
+        Ok(ProxyResponse { status, body })
+    }
+
+    async fn route_collect(
+        &self,
+        engines: &Vec<EngineInfo>,
+        method: Method,
+        api_path: &str,
+        request: Option<&serde_json::Value>,
+    ) -> Result<Vec<ProxyResponse>, actix_web::Error> {
+        let tasks = engines
+            .iter()
+            .map(|engine| self.route_one(engine, method.clone(), api_path, request, false));
+        let responses = join_all(tasks).await;
+        responses
+            .into_iter()
+            .map(|r| r.map_err(actix_web::error::ErrorBadGateway))
+            .collect()
+    }
+
+    async fn generate(
+        &self,
+        api_path: &str,
+        mut req: Box<dyn Bootstrap>,
+    ) -> Result<HttpResponse, actix_web::Error> {
+        let (prefill, decode) = self.strategy_lb.select_pair(&self.client).await;
```

Comment:
question: it seems each `generate` will call `select_pair`, and thus one call to select_pd_pair_po2, and thus four calls to get_load_from_engine, and thus four http requests. Maybe we can reduce that a bit?